RecoEgamma/EgammaElectronProducers/plugins/LowPtGsfElectronSCProducer.cc

```diff
@@ -0,0 +1,195 @@
+	tmpClusters.push_back(best_kf);
+	best_seed = best_kf; // Use KF as seed
+	reco::CaloClusterPtr ptr(edm::refToPtr(best_kf));
+	caloClusters->push_back(*ptr); // Copy CaloCluster 
+        caloClustersMap[ptr] = caloClusters->size() - 1;
+      }
+    }
+
+    // Now make the SuperCluster
+    if ( !best_seed.isNull() ) {
+
+      float posX=0.,posY=0.,posZ=0.;
+      float scEnergy=0.;
+      for ( const auto clus : tmpClusters ) {
+	scEnergy+=clus->correctedEnergy();
+	posX+=clus->correctedEnergy()*clus->position().X();
+	posY+=clus->correctedEnergy()*clus->position().Y();
+	posZ+=clus->correctedEnergy()*clus->position().Z();
+      }
+      posX/=scEnergy;
+      posY/=scEnergy;
+      posZ/=scEnergy;
+      reco::SuperCluster sc(scEnergy,math::XYZPoint(posX,posY,posZ));
+      sc.setCorrectedEnergy(scEnergy);
+      sc.setSeed(edm::refToPtr(best_seed));
+      std::vector<const reco::PFCluster*> barePtrs;
+      for ( const auto clus : tmpClusters ) {
+	sc.addCluster(edm::refToPtr(clus));
+	barePtrs.push_back(&*clus);
+      }
+      PFClusterWidthAlgo pfwidth(barePtrs);
+      sc.setEtaWidth(pfwidth.pflowEtaWidth());
+      sc.setPhiWidth(pfwidth.pflowPhiWidth());
+      sc.rawEnergy(); // Cache the value of raw energy
+
+      // Store new SuperCluster 
+      superClusters->push_back(sc);
+      
+      // Populate ValueMap container
+      superClustersValueMap.push_back( reco::SuperClusterRef(superClustersRefProd,igsfpf) );
+    } else {
+      superClustersValueMap.push_back( reco::SuperClusterRef(superClustersRefProd.id()) );
+    }
+  }
+
+  // Put CaloClusters in event first
+  const edm::OrphanHandle<reco::CaloClusterCollection>& caloClustersH = event.put(std::move(caloClusters));
+
+  // Update CaloClusterRefs in SuperClusters
+  for ( auto sc : *superClusters ) {
+    sc.setSeed( reco::CaloClusterPtr( caloClustersH, caloClustersMap[sc.seed()] ) );
+    reco::CaloClusterPtrVector clusters;
+    for ( auto clu : sc.clusters() ) {
+      clusters.push_back( reco::CaloClusterPtr( caloClustersH, caloClustersMap[clu] ) );
+    }
+    sc.setClusters(clusters);
+  }
+
+  // Put SuperClusters in event
+  event.put(std::move(superClusters));
+
+  // Put ValueMap<SuperClusterRef> in event
+  auto ptr = std::make_unique< edm::ValueMap<reco::SuperClusterRef> >( edm::ValueMap<reco::SuperClusterRef>() );
+  edm::ValueMap<reco::SuperClusterRef>::Filler filler(*ptr);
+  filler.insert(gsfPfRecTracks, superClustersValueMap.begin(), superClustersValueMap.end());
+  filler.fill();
+  event.put(std::move(ptr));
+
+}
+
+reco::PFClusterRef LowPtGsfElectronSCProducer::closestCluster( const reco::PFTrajectoryPoint& point,
+							       const edm::Handle<reco::PFClusterCollection>& clusters,
+							       std::vector<int>& matched ) {
+  reco::PFClusterRef closest;
+  if ( point.isValid() ) {
+    float dr2min = 1.e6;
+    for ( size_t ii = 0; ii < clusters->size(); ++ii ) {
+      if ( std::find( matched.begin(), matched.end(), ii ) == matched.end() ) {
```

Comment:
This would be faster with a std::set, also I am not sure is the correct strategy, as it gives priority to the first matching rather than the better

Reply:
Maybe a better approach could be this. (to be run on all the extrapolated points at once)
```C++
std::vector<reco::PFClusterRef> LowPtGsfElectronSCProducer::closestClusters( const std::vector<reco::PFTrajectoryPoint>& points,
                                                               const edm::Handle<reco::PFClusterCollection>& clusters) {
  // temp var to keep the best DR(point, cluster) and the cluster index for each point
  float dr2min[points.size()];
  int cluster_idx[points.size()];
  for(size_t jj=0; jj < points.size(); ++jj) {
    dr2min[jj] = 1.e6;
    cluster_idx[jj] = -1;
  }

  // loop over clusters
  for(size_t ii = 0; ii < clusters->size(); ++ii) {
    // find the point closest to the cluster and store the DR
    float cluster_dr2min = 1.e6;
    int point_idx = -1;
    for(size_t jj=0; jj < points.size(); ++jj) {
      if( points[jj].isValid() ) {
	float dr2 = reco::deltaR2( clusters->at(ii), points[jj].positionREP() );
	if ( dr2 < cluster_dr2min ) {
          cluster_dr2min = dr2;
          point_idx = jj;
	}
      }
    }

    // check if the cluster found is closer than the current best. if so, store it
    if (point_idx >= 0 && cluster_dr2min < dr2min[point_idx]) {
      dr2min[point_idx] = cluster_dr2min;
      cluster_idx[point_idx] = ii;
    }
  }

  //convert the indices into a cluster vector
  std::vector<reco::PFClusterRef> ret_val;
  for(size_t jj=0; jj < points.size(); ++jj) {
    if(cluster_idx[jj] >= 0) {
      ret_val.emplace_back(clusters, cluster_idx[jj]);
    }
  }
  return ret_val;
}
```